File: CiaiControllerSDK/Communication/SerialCommunication.cs
using System;
using System.Collections.Generic;
using System.IO.Ports;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using CiaiControllerSDK.Interfaces;
using CiaiControllerSDK.Logging;

namespace CiaiControllerSDK.Communication
{
    /// <summary>
    /// 串口通信实现。所有读写通过同一事务信号量，避免多个接口交叉收发。
    /// </summary>
    public class SerialCommunication : IFramedCommunication, IDisposable
    {
        private readonly string _portName;
        private readonly int _baudRate;
        private readonly Parity _parity;
        private readonly int _dataBits;
        private readonly StopBits _stopBits;
        private readonly int _readTimeout;
        private readonly int _writeTimeout;
        private readonly Encoding _encoding;
        private readonly Handshake _handshake;
        private readonly bool _dtrEnable;
        private readonly bool _rtsEnable;
        private readonly bool _discardInputBeforeWrite;
        private readonly SemaphoreSlim _accessSemaphore = new(1, 1);
        private readonly ILogger _logger;
        private SerialPort _serialPort;
        private bool _disposed;

        public bool IsConnected => !_disposed && _serialPort?.IsOpen == true;
        public string PortName => _portName;
        public int BaudRate => _baudRate;

        public SerialCommunication(string portName, int baudRate = 9600,
            Parity parity = Parity.None, int dataBits = 8, StopBits stopBits = StopBits.One,
            int timeout = 5000, Encoding encoding = null)
            : this(portName, baudRate, parity, dataBits, stopBits, timeout, timeout, encoding)
        {
        }

        public SerialCommunication(string portName, int baudRate, Parity parity, int dataBits,
            StopBits stopBits, int readTimeout, int writeTimeout, Encoding encoding = null,
            Handshake handshake = Handshake.None, bool dtrEnable = false, bool rtsEnable = false,
            bool discardInputBeforeWrite = false)
        {
            if (string.IsNullOrWhiteSpace(portName))
                throw new ArgumentException("串口名称不能为空", nameof(portName));
            if (baudRate <= 0)
                throw new ArgumentOutOfRangeException(nameof(baudRate));
            if (dataBits < 5 || dataBits > 8)
                throw new ArgumentOutOfRangeException(nameof(dataBits), "串口数据位必须在5到8之间");
            if (stopBits == StopBits.None)
                throw new ArgumentOutOfRangeException(nameof(stopBits), "串口停止位不能为None");
            if (readTimeout <= 0 || writeTimeout <= 0)
                throw new ArgumentOutOfRangeException(nameof(readTimeout), "串口超时必须大于0");

            _portName = portName;
            _baudRate = baudRate;
            _parity = parity;
            _dataBits = dataBits;
            _stopBits = stopBits;
            _readTimeout = readTimeout;
            _writeTimeout = writeTimeout;
            _encoding = encoding ?? Encoding.UTF8;
            _handshake = handshake;
            _dtrEnable = dtrEnable;
            _rtsEnable = rtsEnable;
            _discardInputBeforeWrite = discardInputBeforeWrite;
            _logger = LoggerProvider.CreateLogger<SerialCommunication>();
        }

        public async Task<bool> ConnectAsync()
        {
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(CancellationToken.None))
                return false;
            try
            {
                DisconnectCore();
                _logger.LogDebug(
                    "正在打开串口: {PortName}, {BaudRate}/{DataBits}/{Parity}/{StopBits}",
                    _portName, _baudRate, _dataBits, _parity, _stopBits);

                _serialPort = new SerialPort(_portName, _baudRate, _parity, _dataBits, _stopBits)
                {
                    ReadTimeout = _readTimeout,
                    WriteTimeout = _writeTimeout,
                    Encoding = _encoding,
                    Handshake = _handshake,
                    DtrEnable = _dtrEnable,
                    RtsEnable = _rtsEnable
                };
                _serialPort.Open();
                _logger.LogInformation("串口打开成功: {PortName}", _portName);
                return true;
            }
            catch (Exception ex)
            {
                DisconnectCore();
                _logger.LogError(ex, "打开串口失败: {PortName}", _portName);
                return false;
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        public async Task DisconnectAsync()
        {
            if (_disposed && _serialPort == null)
                return;
            await _accessSemaphore.WaitAsync();
            try
            {
                DisconnectCore();
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        public Task<bool> SendAsync(byte[] data) => SendAsync(data, CancellationToken.None);

        public async Task<bool> SendAsync(byte[] data, CancellationToken cancellationToken)
        {
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return false;
            try
            {
                return await SendCoreAsync(data, cancellationToken);
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        public Task<byte[]> ReceiveAsync() => ReceiveAsync(CancellationToken.None);

        public async Task<byte[]> ReceiveAsync(CancellationToken cancellationToken)
        {
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await ReceiveCoreAsync(cancellationToken);
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        public Task<byte[]> SendAndReceiveAsync(byte[] data) =>
            SendAndReceiveAsync(data, CancellationToken.None);

        public async Task<byte[]> SendAndReceiveAsync(byte[] data, CancellationToken cancellationToken)
        {
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await SendCoreAsync(data, cancellationToken)
                    ? await ReceiveCoreAsync(cancellationToken)
                    : null;
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        public Task<bool> SendCommandAsync(string command)
        {
            if (command == null)
                throw new ArgumentNullException(nameof(command));
            return SendAsync(_encoding.GetBytes(command));
        }

        public async Task<string> ReceiveResponseAsync()
        {
            var data = await ReceiveAsync();
            return data == null ? null : _encoding.GetString(data);
        }

        public async Task<string> SendAndReceiveStringAsync(string command)
        {
            if (command == null)
                throw new ArgumentNullException(nameof(command));
            var data = await SendAndReceiveAsync(_encoding.GetBytes(command));
            return data == null ? null : _encoding.GetString(data);
        }

        /// <summary>
        /// 循环读取，直到遇到帧尾字节或超时。整个帧读取期间保持独占访问。
        /// </summary>
        public async Task<byte[]> ReadResponseAsync(int frameEndByte, int timeoutMs,
            CancellationToken cancellationToken = default)
        {
            if (timeoutMs <= 0)
                throw new ArgumentOutOfRangeException(nameof(timeoutMs));
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await ReadUntilCoreAsync((byte)(frameEndByte & 0xFF), timeoutMs, 1024 * 1024,
                    cancellationToken);
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        /// <summary>
        /// 读取指定长度的串口帧。
        /// </summary>
        public async Task<byte[]> ReadExactAsync(int length, int? timeoutMs = null,
            CancellationToken cancellationToken = default)
        {
            if (length <= 0)
                throw new ArgumentOutOfRangeException(nameof(length));
            var timeout = timeoutMs ?? _readTimeout;
            if (timeout <= 0)
                throw new ArgumentOutOfRangeException(nameof(timeoutMs));
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await ReadExactCoreAsync(length, timeout, cancellationToken);
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        async Task<byte[]> IFramedCommunication.ReadExactAsync(int length,
            CancellationToken cancellationToken) =>
            await ReadExactAsync(length, null, cancellationToken);

        public async Task<byte[]> ReadUntilAsync(byte endByte, int maxLength = 1024 * 1024,
            CancellationToken cancellationToken = default)
        {
            if (maxLength <= 0)
                throw new ArgumentOutOfRangeException(nameof(maxLength));
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await ReadUntilCoreAsync(endByte, _readTimeout, maxLength, cancellationToken);
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        /// <summary>
        /// 原子发送并按结束字节读取响应。
        /// </summary>
        public async Task<byte[]> SendAndReadUntilAsync(byte[] data, int frameEndByte,
            int? timeoutMs = null, CancellationToken cancellationToken = default)
        {
            var timeout = timeoutMs ?? _readTimeout;
            if (timeout <= 0)
                throw new ArgumentOutOfRangeException(nameof(timeoutMs));
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await SendCoreAsync(data, cancellationToken)
                    ? await ReadUntilCoreAsync((byte)(frameEndByte & 0xFF), timeout, 1024 * 1024,
                        cancellationToken)
                    : null;
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        /// <summary>
        /// 原子发送并读取指定长度响应。
        /// </summary>
        public async Task<byte[]> SendAndReadExactAsync(byte[] data, int length,
            int? timeoutMs = null, CancellationToken cancellationToken = default)
        {
            if (length <= 0)
                throw new ArgumentOutOfRangeException(nameof(length));
            var timeout = timeoutMs ?? _readTimeout;
            if (timeout <= 0)
                throw new ArgumentOutOfRangeException(nameof(timeoutMs));
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await SendCoreAsync(data, cancellationToken)
                    ? await ReadExactCoreAsync(length, timeout, cancellationToken)
                    : null;
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        async Task<byte[]> IFramedCommunication.SendAndReadExactAsync(byte[] data, int length,
            CancellationToken cancellationToken) =>
            await SendAndReadExactAsync(data, length, null, cancellationToken);

        public async Task<byte[]> SendAndReadUntilAsync(byte[] data, byte endByte,
            int maxLength = 1024 * 1024, CancellationToken cancellationToken = default)
        {
            if (maxLength <= 0)
                throw new ArgumentOutOfRangeException(nameof(maxLength));
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken))
                return null;
            try
            {
                return await SendCoreAsync(data, cancellationToken)
                    ? await ReadUntilCoreAsync(endByte, _readTimeout, maxLength, cancellationToken)
                    : null;
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        public async Task<byte[]> ReadUntilAsync(byte[] delimiter, int maxLength = 1024 * 1024,
            CancellationToken cancellationToken = default)
        {
            ValidateDelimiter(delimiter, maxLength);
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken)) return null;
            try { return await ReadUntilCoreAsync(delimiter, _readTimeout, maxLength, cancellationToken); }
            finally { _accessSemaphore.Release(); }
        }

        public async Task<byte[]> SendAndReadUntilAsync(byte[] data, byte[] delimiter,
            int maxLength = 1024 * 1024, CancellationToken cancellationToken = default)
        {
            ValidateDelimiter(delimiter, maxLength);
            ThrowIfDisposed();
            if (!await WaitForAccessAsync(cancellationToken)) return null;
            try { return await SendCoreAsync(data, cancellationToken)
                    ? await ReadUntilCoreAsync(delimiter, _readTimeout, maxLength, cancellationToken) : null; }
            finally { _accessSemaphore.Release(); }
        }

        public void ClearBuffer()
        {
            ThrowIfDisposed();
            _accessSemaphore.Wait();
            try
            {
                if (IsConnected)
                    _serialPort.DiscardInBuffer();
            }
            finally
            {
                _accessSemaphore.Release();
            }
        }

        private async Task<bool> SendCoreAsync(byte[] data, CancellationToken cancellationToken)
        {
            if (data == null)
                throw new ArgumentNullException(nameof(data));
            if (!IsConnected)
            {
                _logger.LogWarning("发送失败: 串口未打开");
                return false;
            }

            try
            {
                if (_discardInputBeforeWrite)
                    _serialPort.DiscardInBuffer();
                await Task.Run(() => _serialPort.Write(data, 0, data.Length), cancellationToken);
                _logger.LogDebug("串口发送数据: {Length}字节", data.Length);
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "串口发送数据失败: {Length}字节", data.Length);
                return false;
            }
        }

        private async Task<byte[]> ReceiveCoreAsync(CancellationToken cancellationToken)
        {
            if (!IsConnected)
            {
                _logger.LogWarning("接收失败: 串口未打开");
                return null;
            }

            try
            {
                return await Task.Run(() =>
                {
                    var buffer = new byte[4096];
                    var bytesRead = _serialPort.Read(buffer, 0, buffer.Length);
                    if (bytesRead <= 0)
                        return null;
                    var result = new byte[bytesRead];
                    Array.Copy(buffer, result, bytesRead);
                    return result;
                }, cancellationToken);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "串口接收数据失败");
                return null;
            }
        }

        private async Task<byte[]> ReadUntilCoreAsync(byte endByte, int timeoutMs, int maxLength,
            CancellationToken cancellationToken)
        {
            if (!IsConnected)
                return null;
            var result = new List<byte>();
            var deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
            while (DateTime.UtcNow < deadline)
            {
                var value = ReadAvailableByte();
                if (value >= 0)
                {
                    result.Add((byte)value);
                    if ((byte)value == endByte)
                        return result.ToArray();
                    if (result.Count >= maxLength)
                        throw new InvalidOperationException($"串口帧超过最大长度: {maxLength}");
                }
                else if (!await DelayForDataAsync(cancellationToken))
                {
                    return null;
                }
            }

            _logger.LogWarning("串口读取帧超时: {Timeout}ms, 已收到 {Length} 字节", timeoutMs, result.Count);
            return null;
        }

        private async Task<byte[]> ReadExactCoreAsync(int length, int timeoutMs,
            CancellationToken cancellationToken)
        {
            if (!IsConnected)
                return null;
            var result = new List<byte>(length);
            var deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
            while (result.Count < length && DateTime.UtcNow < deadline)
            {
                var value = ReadAvailableByte();
                if (value >= 0)
                    result.Add((byte)value);
                else if (!await DelayForDataAsync(cancellationToken))
                    return null;
            }

            return result.Count == length ? result.ToArray() : null;
        }

        private async Task<byte[]> ReadUntilCoreAsync(byte[] delimiter, int timeoutMs, int maxLength,
            CancellationToken cancellationToken)
        {
            var result = new List<byte>();
            var deadline = DateTime.UtcNow.AddMilliseconds(timeoutMs);
            while (DateTime.UtcNow < deadline && result.Count < maxLength)
            {
                var value = ReadAvailableByte();
                if (value >= 0)
                {
                    result.Add((byte)value);
                    if (result.Count >= delimiter.Length)
                    {
                        var match = true;
                        for (var i = 0; i < delimiter.Length; i++)
                            if (result[result.Count - delimiter.Length + i] != delimiter[i]) { match = false; break; }
                        if (match) return result.ToArray();
                    }
                }
                else if (!await DelayForDataAsync(cancellationToken)) return null;
            }
            return null;
        }

        private static void ValidateDelimiter(byte[] delimiter, int maxLength)
        {
            if (delimiter == null || delimiter.Length == 0) throw new ArgumentException("帧分隔符不能为空", nameof(delimiter));
            if (maxLength < delimiter.Length) throw new ArgumentOutOfRangeException(nameof(maxLength));
        }

        private int ReadAvailableByte()
        {
            return _serialPort?.IsOpen == true && _serialPort.BytesToRead > 0
                ? _serialPort.ReadByte()
                : -1;
        }

        private static async Task<bool> DelayForDataAsync(CancellationToken cancellationToken)
        {
            try
            {
                await Task.Delay(20, cancellationToken);
                return true;
            }
            catch (OperationCanceledException)
            {
                return false;
            }
        }

        private void DisconnectCore()
        {
            if (_serialPort != null)
                _logger.LogInformation("关闭串口: {PortName}", _portName);
            try
            {
                _serialPort?.Close();
            }
            finally
            {
                _serialPort?.Dispose();
                _serialPort = null;
            }
        }

        private async Task<bool> WaitForAccessAsync(CancellationToken cancellationToken)
        {
            try
            {
                await _accessSemaphore.WaitAsync(cancellationToken);
                return true;
            }
            catch (OperationCanceledException)
            {
                return false;
            }
        }

        private void ThrowIfDisposed()
        {
            if (_disposed)
                throw new ObjectDisposedException(nameof(SerialCommunication));
        }

        public bool Connect() => ConnectAsync().GetAwaiter().GetResult();
        public void Disconnect() => DisconnectAsync().GetAwaiter().GetResult();
        public bool Send(byte[] data) => SendAsync(data).GetAwaiter().GetResult();
        public byte[] Receive() => ReceiveAsync().GetAwaiter().GetResult();
        public byte[] SendAndReceive(byte[] data) => SendAndReceiveAsync(data).GetAwaiter().GetResult();
        public bool SendCommand(string command) => SendCommandAsync(command).GetAwaiter().GetResult();
        public string ReceiveResponse() => ReceiveResponseAsync().GetAwaiter().GetResult();
        public string SendAndReceiveString(string command) =>
            SendAndReceiveStringAsync(command).GetAwaiter().GetResult();
        public byte[] ReadResponse(int frameEndByte, int timeoutMs) =>
            ReadResponseAsync(frameEndByte, timeoutMs).GetAwaiter().GetResult();
        public byte[] ReadExact(int length, int? timeoutMs = null) =>
            ReadExactAsync(length, timeoutMs).GetAwaiter().GetResult();
        public byte[] SendAndReadUntil(byte[] data, int frameEndByte, int? timeoutMs = null) =>
            SendAndReadUntilAsync(data, frameEndByte, timeoutMs).GetAwaiter().GetResult();
        public byte[] SendAndReadExact(byte[] data, int length, int? timeoutMs = null) =>
            SendAndReadExactAsync(data, length, timeoutMs).GetAwaiter().GetResult();

        public void Dispose()
        {
            if (_disposed)
                return;
            Disconnect();
            _disposed = true;
            _accessSemaphore.Dispose();
        }
    }
}

File: CiaiControllerSDKForJava/src/main/java/com/ciai/controller/sdk/communication/SerialCommunication.java
package com.ciai.controller.sdk.communication;

import com.ciai.controller.sdk.interface_.ICommunication;
import com.ciai.controller.sdk.interface_.IFramedCommunication;
import com.fazecast.jSerialComm.SerialPort;
import com.ciai.controller.sdk.logging.LoggerProvider;
import org.slf4j.Logger;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.Charset;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.ReentrantLock;

/**
 * 串口通信实现 (基于 jSerialComm)
 */
public class SerialCommunication implements IFramedCommunication, AutoCloseable {

    private static final Logger logger = LoggerProvider.createLogger(SerialCommunication.class);

    private final String portName;
    private final int baudRate;
    private final int dataBits;
    private final int stopBits;
    private final int parity;
    private final int readTimeout;
    private final int writeTimeout;
    private final Charset encoding;
    private final int flowControl;
    private final boolean dtrEnable;
    private final boolean rtsEnable;
    private final boolean discardInputBeforeWrite;
    private final ReentrantLock transactionLock = new ReentrantLock(true);

    private SerialPort serialPort; // jSerialComm serial port object
    private InputStream inputStream;
    private OutputStream outputStream;
    private final ExecutorService executor = Executors.newFixedThreadPool(2);
    private boolean connected = false;

    // Parity constants
    public static final int PARITY_NONE = 0;
    public static final int PARITY_ODD = 1;
    public static final int PARITY_EVEN = 2;
    public static final int PARITY_MARK = 3;
    public static final int PARITY_SPACE = 4;

    // Stop bits constants
    public static final int STOPBITS_ONE = 1;
    public static final int STOPBITS_ONE_POINT_FIVE = 3;
    public static final int STOPBITS_TWO = 2;

    public SerialCommunication(String portName) {
        this(portName, 9600);
    }

    public SerialCommunication(String portName, int baudRate) {
        this(portName, baudRate, 8, STOPBITS_ONE, PARITY_NONE, 5000);
    }

    public SerialCommunication(String portName, int baudRate, int dataBits, int stopBits, int parity, int timeout) {
        this(portName, baudRate, dataBits, stopBits, parity, timeout, timeout, Charset.forName("UTF-8"));
    }

    public SerialCommunication(String portName, int baudRate, int dataBits, int stopBits, int parity,
                               int readTimeout, int writeTimeout, Charset encoding) {
        this(portName, baudRate, dataBits, stopBits, parity, readTimeout, writeTimeout, encoding,
                SerialPort.FLOW_CONTROL_DISABLED, false, false, false);
    }

    public SerialCommunication(String portName, int baudRate, int dataBits, int stopBits, int parity,
                               int readTimeout, int writeTimeout, Charset encoding, int flowControl,
                               boolean dtrEnable, boolean rtsEnable, boolean discardInputBeforeWrite) {
        if (portName == null || portName.trim().isEmpty()) {
            throw new IllegalArgumentException("Serial port name is required");
        }
        if (baudRate <= 0 || dataBits < 5 || dataBits > 8) {
            throw new IllegalArgumentException("Invalid serial port parameters");
        }
        if (readTimeout <= 0 || writeTimeout <= 0) {
            throw new IllegalArgumentException("Serial timeouts must be greater than zero");
        }
        this.portName = portName;
        this.baudRate = baudRate;
        this.dataBits = dataBits;
        this.stopBits = stopBits;
        this.parity = parity;
        this.readTimeout = readTimeout;
        this.writeTimeout = writeTimeout;
        this.encoding = encoding == null ? Charset.forName("UTF-8") : encoding;
        this.flowControl = flowControl;
        this.dtrEnable = dtrEnable;
        this.rtsEnable = rtsEnable;
        this.discardInputBeforeWrite = discardInputBeforeWrite;
    }

    @Override
    public boolean isConnected() {
        return connected && inputStream != null && outputStream != null;
    }

    @Override
    public CompletableFuture<Boolean> connectAsync() {
        return CompletableFuture.supplyAsync(this::connect, executor);
    }

    @Override
    public CompletableFuture<Void> disconnectAsync() {
        return CompletableFuture.runAsync(this::disconnect, executor);
    }

    @Override
    public CompletableFuture<Boolean> sendAsync(byte[] data) {
        return CompletableFuture.supplyAsync(() -> send(data), executor);
    }

    @Override
    public CompletableFuture<byte[]> receiveAsync() {
        return CompletableFuture.supplyAsync(this::receive, executor);
    }

    @Override
    public CompletableFuture<byte[]> sendAndReceiveAsync(byte[] data) {
        return CompletableFuture.supplyAsync(() -> sendAndReceive(data), executor);
    }

    @Override
    public CompletableFuture<byte[]> readExactAsync(int length) {
        return CompletableFuture.supplyAsync(() -> readExact(length), executor);
    }

    @Override
    public CompletableFuture<byte[]> readUntilAsync(byte endByte, int maxLength) {
        return CompletableFuture.supplyAsync(() -> readUntil(endByte, maxLength), executor);
    }

    @Override
    public CompletableFuture<byte[]> sendAndReadExactAsync(byte[] data, int length) {
        return CompletableFuture.supplyAsync(() -> sendAndReadExact(data, length), executor);
    }

    @Override
    public CompletableFuture<byte[]> sendAndReadUntilAsync(byte[] data, byte endByte, int maxLength) {
        return CompletableFuture.supplyAsync(
                () -> sendAndReadUntil(data, endByte, maxLength), executor);
    }

    @Override public CompletableFuture<byte[]> readUntilAsync(byte[] delimiter, int maxLength) {
        return CompletableFuture.supplyAsync(() -> readUntil(delimiter, maxLength), executor);
    }
    @Override public CompletableFuture<byte[]> sendAndReadUntilAsync(byte[] data, byte[] delimiter, int maxLength) {
        return CompletableFuture.supplyAsync(() -> sendAndReadUntil(data, delimiter, maxLength), executor);
    }

    @Override
    public boolean connect() {
        transactionLock.lock();
        logger.info("Attempting to connect to serial port: {} at {} baud, {} data bits, {} stop bits, parity {}",
                portName, baudRate, dataBits, stopBits, parity);

        try {
            disconnectCore();
            serialPort = SerialPort.getCommPort(portName);
            serialPort.setComPortParameters(baudRate, dataBits, stopBits, parity);
            serialPort.setFlowControl(flowControl);
            serialPort.setComPortTimeouts(
                    SerialPort.TIMEOUT_READ_SEMI_BLOCKING | SerialPort.TIMEOUT_WRITE_BLOCKING,
                    readTimeout, writeTimeout);

            if (serialPort.openPort()) {
                if (dtrEnable) serialPort.setDTR(); else serialPort.clearDTR();
                if (rtsEnable) serialPort.setRTS(); else serialPort.clearRTS();
                inputStream = serialPort.getInputStream();
                outputStream = serialPort.getOutputStream();
                connected = true;
                logger.info("Serial port {} connected successfully", portName);
                return true;
            } else {
                logger.error("Failed to open serial port {}", portName);
                return false;
            }
        } catch (Exception e) {
            disconnectCore();
            logger.error("Serial port connection failed: {}", portName, e);
            return false;
        } finally {
            transactionLock.unlock();
        }
    }

    @Override
    public void disconnect() {
        transactionLock.lock();
        try {
            disconnectCore();
            logger.info("Serial port {} disconnected", portName);
        } finally {
            transactionLock.unlock();
        }
    }

    @Override
    public boolean send(byte[] data) {
        transactionLock.lock();
        try {
            return sendCore(data);
        } finally {
            transactionLock.unlock();
        }
    }

    private boolean sendCore(byte[] data) {
        if (!isConnected() || outputStream == null) {
            logger.error("Serial port not connected, cannot send");
            return false;
        }
        try {
            if (discardInputBeforeWrite) serialPort.flushIOBuffers();
            outputStream.write(data);
            outputStream.flush();
            logger.debug("Serial sent {} bytes", data.length);
            return true;
        } catch (IOException e) {
            logger.error("Serial send error", e);
            return false;
        }
    }

    @Override
    public byte[] receive() {
        transactionLock.lock();
        try {
            return receiveCore();
        } finally {
            transactionLock.unlock();
        }
    }

    private byte[] receiveCore() {
        if (!isConnected() || inputStream == null) {
            logger.error("Serial port not connected, cannot receive");
            return null;
        }
        try {
            byte[] buffer = new byte[4096];
            int bytesRead = inputStream.read(buffer);
            if (bytesRead > 0) {
                byte[] result = new byte[bytesRead];
                System.arraycopy(buffer, 0, result, 0, bytesRead);
                logger.debug("Serial received {} bytes", bytesRead);
                return result;
            }
            return null;
        } catch (IOException e) {
            logger.error("Serial receive error", e);
            return null;
        }
    }

    @Override
    public byte[] sendAndReceive(byte[] data) {
        transactionLock.lock();
        try {
            if (sendCore(data)) {
                return receiveCore();
            }
            return null;
        } finally {
            transactionLock.unlock();
        }
    }

    /**
     * 发送字符串命令
     */
    public CompletableFuture<Boolean> sendCommandAsync(String command) {
        return CompletableFuture.supplyAsync(() -> sendCommand(command), executor);
    }

    /**
     * 发送字符串命令
     */
    public boolean sendCommand(String command) {
        return send(command.getBytes(encoding));
    }

    /**
     * 接收字符串响应
     */
    public CompletableFuture<String> receiveResponseAsync() {
        return CompletableFuture.supplyAsync(this::receiveResponse, executor);
    }

    /**
     * 接收字符串响应
     */
    public String receiveResponse() {
        byte[] data = receive();
        return data != null ? new String(data, encoding) : null;
    }

    /**
     * 发送并接收字符串
     */
    public CompletableFuture<String> sendAndReceiveStringAsync(String command) {
        return CompletableFuture.supplyAsync(() -> sendAndReceiveString(command), executor);
    }

    /**
     * 发送并接收字符串
     */
    public String sendAndReceiveString(String command) {
        transactionLock.lock();
        try {
            byte[] result = sendAndReceive(command.getBytes(encoding));
            return result != null ? new String(result, encoding) : null;
        } finally {
            transactionLock.unlock();
        }
    }

    /**
     * 循环读取串口响应直到帧尾标识或超时（参照原封膜机TCPClient.Read实现）
     * @param frameEndByte 帧尾标识字节(如0xFF)
     * @param timeoutMs 超时毫秒
     * @return 完整响应字节数组，超时返回null
     */
    public byte[] readResponse(int frameEndByte, int timeoutMs) {
        transactionLock.lock();
        try {
            return readResponseCore(frameEndByte, timeoutMs, 1024 * 1024);
        } finally {
            transactionLock.unlock();
        }
    }

    @Override
    public byte[] readExact(int length) {
        transactionLock.lock();
        try {
            return readExactCore(length);
        } finally {
            transactionLock.unlock();
        }
    }

    @Override
    public byte[] readUntil(byte endByte, int maxLength) {
        transactionLock.lock();
        try {
            return readResponseCore(endByte & 0xFF, readTimeout, maxLength);
        } finally {
            transactionLock.unlock();
        }
    }

    @Override
    public byte[] sendAndReadExact(byte[] data, int length) {
        transactionLock.lock();
        try {
            return sendCore(data) ? readExactCore(length) : null;
        } finally {
            transactionLock.unlock();
        }
    }

    @Override
    public byte[] sendAndReadUntil(byte[] data, byte endByte, int maxLength) {
        transactionLock.lock();
        try {
            return sendCore(data)
                    ? readResponseCore(endByte & 0xFF, readTimeout, maxLength)
                    : null;
        } finally {
            transactionLock.unlock();
        }
    }

    @Override public byte[] readUntil(byte[] delimiter, int maxLength) {
        transactionLock.lock(); try { return readUntilCore(delimiter, maxLength); }
        finally { transactionLock.unlock(); }
    }
    @Override public byte[] sendAndReadUntil(byte[] data, byte[] delimiter, int maxLength) {
        transactionLock.lock(); try { return sendCore(data) ? readUntilCore(delimiter, maxLength) : null; }
        finally { transactionLock.unlock(); }
    }

    private byte[] readExactCore(int length) {
        if (length <= 0) {
            throw new IllegalArgumentException("Frame length must be greater than zero");
        }
        if (!isConnected() || inputStream == null) {
            logger.error("Serial port not connected, cannot read frame");
            return null;
        }
        byte[] result = new byte[length];
        int offset = 0;
        try {
            while (offset < length) {
                int count = inputStream.read(result, offset, length - offset);
                if (count < 0) {
                    return null;
                }
                offset += count;
            }
            return result;
        } catch (IOException e) {
            logger.error("Serial exact frame read error", e);
            return null;
        }
    }

    private byte[] readResponseCore(int frameEndByte, int timeoutMs, int maxLength) {
        if (timeoutMs <= 0 || maxLength <= 0) {
            throw new IllegalArgumentException(
                    "Frame timeout and maximum length must be greater than zero");
        }
        if (!isConnected() || inputStream == null) {
            logger.error("Serial port not connected, cannot read response");
            return null;
        }
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        long deadline = System.currentTimeMillis() + timeoutMs;
        try {
            while (System.currentTimeMillis() < deadline && baos.size() < maxLength) {
                int available = inputStream.available();
                if (available > 0) {
                    int b = inputStream.read();
                    baos.write(b);
                    if (b == (frameEndByte & 0xFF)) {
                        byte[] result = baos.toByteArray();
                        logger.debug("Serial readResponse complete: {} bytes", result.length);
                        return result;
                    }
                } else {
                    Thread.sleep(50);
                }
            }
            logger.error("Serial frame incomplete after {}ms or {} bytes; received {} bytes",
                    timeoutMs, maxLength, baos.size());
            return null;
        } catch (Exception e) {
            logger.error("Serial readResponse error", e);
            return null;
        }
    }

    /**
     * 清空串口输入缓冲区（防止残留数据导致响应错乱）
     */
    public void clearBuffer() {
        transactionLock.lock();
        try {
            clearBufferCore();
        } finally {
            transactionLock.unlock();
        }
    }

    private byte[] readUntilCore(byte[] delimiter, int maxLength) {
        if (delimiter == null || delimiter.length == 0 || maxLength < delimiter.length)
            throw new IllegalArgumentException("Invalid frame delimiter or maximum length");
        ByteArrayOutputStream result = new ByteArrayOutputStream();
        long deadline = System.currentTimeMillis() + readTimeout;
        try {
            while (System.currentTimeMillis() < deadline && result.size() < maxLength) {
                if (inputStream.available() == 0) { Thread.sleep(20); continue; }
                result.write(inputStream.read()); byte[] bytes = result.toByteArray();
                boolean match = bytes.length >= delimiter.length;
                for (int i = 0; match && i < delimiter.length; i++)
                    match = bytes[bytes.length - delimiter.length + i] == delimiter[i];
                if (match) return bytes;
            }
            return null;
        } catch (Exception e) { logger.error("Serial delimited frame read error", e); return null; }
    }

    private void clearBufferCore() {
        if (!isConnected() || inputStream == null) {
            return;
        }
        try {
            int drained = 0;
            while (inputStream.available() > 0) {
                inputStream.read();
                drained++;
            }
            if (drained > 0) {
                logger.debug("Serial cleared {} bytes from input buffer", drained);
            }
        } catch (IOException e) {
            logger.warn("Serial clearBuffer error", e);
        }
    }

    private void disconnectCore() {
        try {
            if (inputStream != null) inputStream.close();
        } catch (IOException e) {
            logger.debug("Serial input close error", e);
        } finally {
            inputStream = null;
        }
        try {
            if (outputStream != null) outputStream.close();
        } catch (IOException e) {
            logger.debug("Serial output close error", e);
        } finally {
            outputStream = null;
        }
        if (serialPort != null) {
            serialPort.closePort();
            serialPort = null;
        }
        connected = false;
    }

    @Override
    public void close() {
        disconnect();
        executor.shutdown();
    }
}
